Add kubernetes v4 blog post (#3163)
pulumi/pulumi-hugo: themes/default/content/blog/kubernetes-v4-shared-resource-management-with-server-side-apply/index.md

---

> title: "Kubernetes 4.0: Shared Resource Management With Server-Side Apply"

Reviewer: ```suggestion
title: "Kubernetes Native Provider 4.0: Build on Kubernetes Collaboratively and Safely with Shared Resource Management With Server-Side Apply"
```

Reviewer: suggested edit - suggest bringing in the benefit of SSA into the title somehow since this is the main update for this version. feel free to edit a bit

Reviewer 2: Added my own proposal in a new commit which fits in 60 characters.  @gthuang if you have additional suggested changes that fit in 60 characters, feel free to suggest them.

date: 2023-07-19T09:00:00-04:00

meta_desc: The Pulumi Kubernetes Provider 4.0; Shared Resource Management with Server-Side Apply

meta_image: meta.png

authors:
    - levi-blackstone
    - luke-hoban
    - monica-rodriguez

tags:
    - kubernetes
    - providers

---

> Since the very earliest days of the Pulumi project, Kubernetes has been a core part of the Pulumi platform.  The initial Pulumi Kubernetes provider supported the entire API surface area of the Kubernetes platform, derived directly and automatically from the Kubernetes OpenAPI specifications.  Since then, we have offered day one support for every new Kubernetes version, added support for Helm, YAML, Kustomize and CRDs, added tools for converting to Pulumi (kube2pulumi and crd2pulumi) and delivered the Pulumi Kubernetes Operator.  During that same time, Kubernetes usage has continued to expand within the ecosystem and within Pulumi users, with the Kubernetes provider growing from the fourth most used to the second most used provider on the platform.

Reviewer: ```suggestion
Since the very earliest days of the Pulumi project, Kubernetes has been a core part of the Pulumi platform and has grown from the fourth most used to the second most used provider on the platform. Pulumi simplifies Kubernetes management by allowing you to express your K8s resources with familiar programming languages. Pulumi users love the Kubernetes Native Provider because it provides the most complete and up-to-date coverage of the Kubernetes platform and it natively supports K8s constructs and best practices like CI/CD. The initial Pulumi Kubernetes provider supported the entire API surface area of the Kubernetes platform, derived directly and automatically from the Kubernetes OpenAPI specifications.  Since then, we have offered day-one support for every new Kubernetes version, added support for Helm, YAML, Kustomize and CRDs, added tools for converting to Pulumi (kube2pulumi and crd2pulumi) and delivered the Pulumi Kubernetes Operator.
```

Reviewer: rewrote this a bit to position the usage stats upfront (those are most impressive), then recap why Kubernetes Provider is unique and differentiated for our users. i think this will be more impactful while also educating anyone who's not familiar with the provider yet

Reviewer 2: I took a few of these changes into the latest commit, but left some of it as-is.  Open to additional changes if still needed - though I prefer to not lead with the usage stat in this particular case.

Today, we are excited to release the latest major version of our Kubernetes provider - Pulumi Kubernetes 4.0.

<!--more-->

> Pulumi Kubernetes 4.0 uses Kubernetes Server-Side Apply by default, supports upserting on all resources, and enables new Patch resources for every resource type in the provider.  It also brings improved diffs, removes dependence on kubectl annotations, and provides simpler access to outputs in the Python, Go and Java SDKs, and with any Kubernetes cluster version >= 1.13 and in any self-hosted or managed Kubernetes cluster.

Reviewer 3: ```suggestion
Pulumi Kubernetes 4.0 uses Kubernetes Server-Side Apply by default, supports upserting on all resources, and enables new Patch resources for every resource type in the provider.  It also brings improved diffs, removes dependence on kubectl annotations, and provides simpler access to outputs in the Python, Go and Java SDKs,  supporting any self-hosted or managed Kubernetes cluster with version 1.13 or newer.
```

> Together, these additions make Pulumi’s support for Kuberntes even more Kubernetes-native, allowing the best of both worlds between Kubernetes’ flexible resource model and Pulumi’s robust desired state infrastructure as code model, all while using your favorite languages (TypeScript, Python, Go, .NET and Java) to define and manage your Kubernetes Infrastructure.  Pulumi programs that need to manage Kubernetes resources that are also modified by other systems (external controllers, managed Kubernetes service control planes, kubectl or other IaC solutions) can faithfully represent the modifications they want to make to existing resources within the Pulumi resource lifecycle.  And Pulumi’s diffs will represent the accurate differences based on the current state of resources in the cluster.

Reviewer 3: ```suggestion
Together, these additions make Pulumi’s support for Kuberntes even more Kubernetes-native, allowing the best of both worlds between Kubernetes’ flexible resource model and Pulumi’s robust desired state infrastructure as code model, all while using your favorite languages (TypeScript, Python, Go, .NET and Java) to define and manage your Kubernetes Infrastructure.  Pulumi programs that need to manage Kubernetes resources that are also modified by other systems (external controllers, managed Kubernetes service control planes, kubectl or other IaC solutions) can faithfully represent the modifications they want to make to existing resources within the Pulumi resource lifecycle.  Pulumi will accurately represent any differences between the current state of resources in the cluster, and the desired state specified by Pulumi IaC.
```

> Note: Users may opt into the previous default Client-side Apply (CSA) mode by setting the `enableServerSideApply` provider configuration to “false”.

> {{< chooser language "typescript,python,go,csharp,java,yaml" >}}

Reviewer 3: This example is for upserting a resource. We should add a heading. It might also be worth adding the Patch example from https://www.pulumi.com/registry/packages/kubernetes/how-to-guides/managing-resources-with-server-side-apply/#patch-a-resource

Reviewer 2: Added a note - but still could potentially be nice to include a second code example.  Though in context, I think it is okay to have only one inlined here - and to leave the rest for the migration guide.
